### ISSUE TITLE
Wallaby.end_session/1 seems to be unsafe

### ISSUE THREAD

**Issue by aaronjensen** [keathley/wallaby#131]
At some point we ended up with: 

```elixir
    on_exit fn ->
      Wallaby.end_session(session)
    end
```

In our AcceptanceCase. I don't know why it was added. I know it's not in the readme, but it is in the [tests of wallaby itself].

I have been tracking down a strange bug wherein our wallaby tests would occasionally fail because they would receive ajax responses intended for another test. Logs actually showed two different tests receiving responses with the exact same request id.

I managed to track it down to that `Wallaby.end_session/1`. Removing it fixed the problem and didn't appear to cause any other issues (again, I'm not sure why we added it.)

When the error happens, it appears that multiple tests are able to simultaneously check out the same server (I verified this by logging the `base_url`:

```
08:56:21.892 test=Elixir.TheLink.SignInTest.Foo13:test redirect to original destination after sign in 13 sandbox=%{owner: #PID<0.644.0>, repo: TheLink.Repo} [info] BASEURL: http://localhost:56044/
08:56:21.916 test=Elixir.TheLink.SignInTest.Foo15:test redirect to original destination after sign in 15 sandbox=%{owner: #PID<0.658.0>, repo: TheLink.Repo} [info] BASEURL: http://localhost:56044/
0
```

My guess is that the server is getting checked back in to `poolboy` twice. Once when the test ends, and once with `end_session/1`. 

I looked at the poolboy code, and, as best as I can tell, this should not be possible. Checkin is only allowed if the monitor ([manual checkin](https://github.com/devinus/poolboy/blob/master/src/poolboy.erl#L149-L150), [DOWN](https://github.com/devinus/poolboy/blob/master/src/poolboy.erl#L227)) is still in place for the original process that checked it out and that pid is removed from the monitors upon checkin ([manual checkin](https://github.com/devinus/poolboy/blob/master/src/poolboy.erl#L152), [DOWN](https://github.com/devinus/poolboy/blob/master/src/poolboy.erl#L229)). I'll admit I'm not that great at reading erlang, so I may be missing something here.

Any way, it seems like there's either a bug in poolboy, or in Wallaby, and at the very least, maybe `end_session/1` should be documented as unsafe? Thoughts?

**Comment by keathley** [comment]
Thanks for the great writeup @aaronjensen. It seems like you shouldn't be able to check in sessions more then once. It might be another race condition somewhere. Can you help me get a scaffolding in place to re-create this? That would be useful for re-creating this issue.

**Comment by aaronjensen** [comment]
@keathley Try cloning this: https://github.com/aaronjensen/wallaby-shared-session-repro

Then:

```bash
$ yarn # or npm i
$ mix deps.get
$ mix test
```

If there is a failure, it happened. The two user agents are printed (the agent the client thinks it is and the server thinks it is.

You may need to run `mix test` a couple times to see it.

**Comment by keathley** [comment]
Awesome thanks! I really appreciate you putting that together 👍

**Comment by hisapy** [comment]
Hey guys,

I'm also experiencing some issues with `end_session`. 

In my case, I `Wallaby.start_session` in a setup callback, and `end_session` in an `on_exit`, but when trying to `visit` in subsequent tests, I get the following error:

```bash
** (MatchError) no match of right hand side value: {:error, {:invalid, "V"}}
     stacktrace:
       (wallaby) lib/wallaby/phantom/driver.ex:337: Wallaby.Phantom.Driver.check_logs!/2
       (wallaby) lib/wallaby/browser.ex:954: Wallaby.Browser.visit/2
```
I followed the stack trace and did some `IO.inspect`s and something that caught my attention is that each time `start_session` is called, apparently, the same session is returned. For example:

```bash
%Wallaby.Session{id: "d4598340-15a6-11e7-a266-5b42e5c3c63e", screenshots: [],
 server: #PID<0.478.0>,
 session_url: "http://localhost:52965/session/d4598340-15a6-11e7-a266-5b42e5c3c63e",
 url: "http://localhost:52965/session/d4598340-15a6-11e7-a266-5b42e5c3c63e"}
```

is returned before and after `end_session` is called.

Shouldn't it be a _different_ session each time `start_session` is called?

_PS.: I'm trying to understand the code inside start_session and `:poolboy` stuff_

**Comment by keathley** [comment]
Thats correct. It should be a different session each time. Can I see how you've structured your setup block?

**Comment by keathley** [comment]
I think I may have a theory on whats going on @aaronjensen. Its related to #174. I was able to force this to fail by reducing the worker pool in your example app to 1. Here's what I suspect is happening:

When a session is checked out poolboy sets up a monitor to the process that checked out the worker (in this case a phantom). When that process dies poolboy checks the worker back in. In this case the process that checked out the process is the test. So the test exits, poolboy sees that te test has ended and it checks the phantom back into the pool. `on_exit` callbacks run in a separate process then the test process. So it seems like whats happening is that the worker is either getting checked back in to poolboy twice, or there is some other race condition with manually checking in sessions. 

I'll see if I can work out a fix in wallaby for this.

**Comment by aaronjensen** [comment]
I did not know that `on_exit` runs in a separate process. In that case, then I would guess you're right. Specifically, the session is getting checked in after another test process has already checked it out. Something like this:

1. Test 1 checks out session 1
2. Test 1 ends
3. Poolboy checks in session 1
4. Test 2 checks out session 1
5. on_exit from Test 1 checks in session 1
6. Test 3 checks out session 1
7. 💥

**Comment by hisapy** [comment]
Hi @keathley 

First, my test case:

```elixir
defmodule Webapp.AcceptanceCase do
  use ExUnit.CaseTemplate

  using do
    quote do
      use Wallaby.DSL

      import Webapp.Router.Helpers
      import Webapp.Factory
      import Webapp.AcceptanceCase

      # The default endpoint for testing
      @endpoint Webapp.Endpoint
    end
  end

  setup tags do
    :ok = Ecto.Adapters.SQL.Sandbox.checkout(Webapp.Repo)

    unless tags[:async] do
      Ecto.Adapters.SQL.Sandbox.mode(Webapp.Repo, {:shared, self()})
    end

    metadata = Phoenix.Ecto.SQL.Sandbox.metadata_for(Webapp.Repo, self())
    {:ok, session} = Wallaby.start_session(metadata: metadata)
    IO.puts "\n <<<< start_session >>>>"
    IO.inspect session # This is where I found out that apparently the same session is used even after end_session
    IO.puts "<<<<<>>>>>"
    {:ok, session: session}
  end

end
```

Then in the test:

```elixir
defmodule Webapp.Features.Admin.SomeTest do
  setup %{session: session}  do
    on_exit fn -> 
      Wallaby.end_session(session)
    end
  end

  test "something", %{session: session} do
    user = login_admin(session)    
    # assert stuff here
  end

  test "other stuff", %{session: session} do 
    user = login_admin(session)
    # assert stuff here
  end
  

  defp login_admin(session) do
    user =
      build(:user)
      |> make_admin()
      |> set_password("12341234")
      |> Repo.insert!

    session
    |> visit("/login") # the second time this is called is where I get the error I reported in my previous comment https://github.com/keathley/wallaby/issues/131#issuecomment-290581536
    |> fill_in(Query.text_field("Email"), with: user.email)
    |> fill_in(Query.text_field("Password"), with: user.password)
    |> click(Query.button("Enter"))
    |> find(Query.css("#welcome"))
    
    user
  end
end
````

**Comment by hisapy** [comment]
### More debug info

I added an `IO.inspect session` in the `on_exit` and got the following results:

```bash
# Remember there are 2 tests 

<<<< start_session >>>>
%Wallaby.Session{id: "fb7b2c80-1614-11e7-b1e6-5376a0c9e7b0", screenshots: [],
 server: #PID<0.484.0>,
 session_url: "http://localhost:41423/session/fb7b2c80-1614-11e7-b1e6-5376a0c9e7b0",
 url: "http://localhost:41423/session/fb7b2c80-1614-11e7-b1e6-5376a0c9e7b0"}
<<<<<>>>>>
<<<< END SESSION! >>>>
%Wallaby.Session{id: "fb7b2c80-1614-11e7-b1e6-5376a0c9e7b0", screenshots: [],
 server: #PID<0.484.0>,
 session_url: "http://localhost:41423/session/fb7b2c80-1614-11e7-b1e6-5376a0c9e7b0",
 url: "http://localhost:41423/session/fb7b2c80-1614-11e7-b1e6-5376a0c9e7b0"}
<<<<<>>>>>

 <<<< start_session >>>>
%Wallaby.Session{id: "0d441760-1615-11e7-b1e6-5376a0c9e7b0", screenshots: [],
 server: #PID<0.484.0>,
 session_url: "http://localhost:41423/session/0d441760-1615-11e7-b1e6-5376a0c9e7b0",
 url: "http://localhost:41423/session/0d441760-1615-11e7-b1e6-5376a0c9e7b0"}
<<<<<>>>>>
<<<< END SESSION! >>>>
%Wallaby.Session{id: "0d441760-1615-11e7-b1e6-5376a0c9e7b0", screenshots: [],
 server: #PID<0.484.0>,
 session_url: "http://localhost:41423/session/0d441760-1615-11e7-b1e6-5376a0c9e7b0",
 url: "http://localhost:41423/session/0d441760-1615-11e7-b1e6-5376a0c9e7b0"}
<<<<<>>>>>
```

**Comment by keathley** [comment]
@aaronjensen I think that its probably something similar to that, but its important to distinguish between "sessions" and "servers". A server is a phantomjs process. A session is a user session running on a specific phantom. Theoretically you could run multiple sessions on a single phantom. We don't do that because phantom doesn't isolate sessions properly (shared cookies, shared localStorage, etc.). I think your theory is still correct. Just replace the word session with server :). I'm only being pedantic about this because I think there is a 2nd issue with not cleaning up sessions properly which is what #174 is addressing. They may be related though.

@hisapy I'm not sure I see an issue in your debug logs. Both tests are using the same server but they have different sessions. It doesn't look like your test is marked as `async: true` so whats probably happening is that your checking out a server, the test finishes, server is checked back in, the next test starts and checks out the same worker. I may be missing something though.

**Comment by hisapy** [comment]
Yeah, it's weird. Anyways, for the second time I try to `visit("/login")` I get error:

```bash
** (MatchError) no match of right hand side value: {:error, {:invalid, "V"}}
     stacktrace:
       (wallaby) lib/wallaby/phantom/driver.ex:337: Wallaby.Phantom.Driver.check_logs!/2
       (wallaby) lib/wallaby/browser.ex:954: Wallaby.Browser.visit/2
```

Do you have any idea about where that "V" is coming from? ... have you seen something like that before?

My phantomjs version is **2.1.1.**

**Comment by keathley** [comment]
Its being thrown from the `Driver`. Probably `make_request/3`. Whats happening is that its getting an error from phantoms webdriver which is then being handled incorrectly. We should update `make_request` and `check_for_response_errors` to better find these issues. If you log that out then you'll see the full error message.

**Comment by hisapy** [comment]
When I 

```elixir
    IO.inspect Wallaby.Phantom.Driver.log(session)
    session
    |> visit("/login")
    ... 
```

I get the same error `{:error, {:invalid, "V"}}` ... Is there any way I can get a _more complete_ log?

**Comment by keathley** [comment]
You would have to log the response from inside of wallaby. The function you'll want to log is `Driver.make_request/3`. If you don't wanna mess with it I'll see if I can get a proper fix in later today.

**Comment by hisapy** [comment]
I'm working on a single "scenario" now so I can wait for your fix ... Thanks a lot in advance ... This is a great lib

**Comment by keathley** [comment]
This should be resolved now that #279 has been merged.
